### ISSUE TITLE
Coolab_etapa3

### ISSUE THREAD

**Issue by pamela09oliveira02-ctrl** [pamela09oliveira02-ctrl/trabalho.ead#6]
import networkx as nx
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt
import numpy as np

# --- Configurações ---
RESTAURANT_NODE = 'RESTAURANTE'
NUM_CLUSTERS = 4 # Número de entregadores/rotas

# --- 1. Modelagem do Grafo ---
def load_graph(df_graph, df_points):
    """Carrega o grafo da cidade a partir do DataFrame."""
    G = nx.Graph()
    
    # Adiciona o nó do restaurante, que é o ponto de partida
    restaurant_row = df_points[df_points['Bairro_Grafo'] == RESTAURANT_NODE].iloc[0]
    G.add_node(RESTAURANT_NODE, pos=(restaurant_row['Coordenada_X'], restaurant_row['Coordenada_Y']))
    
    for _, row in df_graph.iterrows():
        # Usamos 'Tempo_min' como peso principal para o A*
        G.add_edge(row['Origem'], row['Destino'], weight=row['Tempo_min'], distance=row['Distancia_km'])
    
    # Adiciona arestas do restaurante para os nós mais próximos (simplificação)
    # Neste cenário simplificado, o restaurante está no nó 'A' (10, 10)
    G.add_edge(RESTAURANT_NODE, 'A', weight=0, distance=0) # Conexão imediata
    
    return G

# --- 2. Algoritmo A* (Menor Caminho) ---
def a_star_path(G, start_node, end_node):
    """Encontra o caminho mais curto usando o algoritmo A*."""
    try:
        # Heurística (h) = 0, tornando-o equivalente ao Dijkstra (admissível)
        path = nx.astar_path(G, start_node, end_node, weight='weight', heuristic=lambda u, v: 0)
        cost = nx.astar_path_length(G, start_node, end_node, weight='weight', heuristic=lambda u, v: 0)
        return path, cost
    except nx.NetworkXNoPath:
        return None, float('inf')

# --- 3. Agrupamento K-Means ---
def cluster_deliveries(df_points, n_clusters):
    """Aplica o K-Means para agrupamento."""
    
    # Remove o ponto do restaurante para o clustering
    df_deliveries = df_points[df_points['Bairro_Grafo'] != RESTAURANT_NODE].copy()
    
    # Limpeza de dados: remove linhas com NaN, caso existam
    df_deliveries.dropna(subset=['Coordenada_X', 'Coordenada_Y'], inplace=True)
    
    # Prepara os dados para o K-Means (apenas coordenadas X e Y)
    X = df_deliveries[['Coordenada_X', 'Coordenada_Y']].values
    
    # Aplica o K-Means
    kmeans = KMeans(n_clusters=n_clusters, random_state=42, n_init=10)
    df_deliveries['Cluster'] = kmeans.fit_predict(X)
    
    # Adiciona o ponto do restaurante de volta para o processamento de rotas
    df_restaurant = df_points[df_points['Bairro_Grafo'] == RESTAURANT_NODE].copy()
    
    return df_deliveries, df_restaurant, kmeans.cluster_centers_

# --- 4. Otimização de Rotas (TSP Simplificado) ---
def optimize_route(G, start_node, delivery_points_df):
    """
    Otimiza a rota para um cluster de entregas.
    A ordem de visita é simplificada (heurística) para cumprir o desafio.
    """
    
    # Lista de nós de entrega únicos no cluster
    unique_delivery_nodes = delivery_points_df['Bairro_Grafo'].unique().tolist()
    
    # Calcula o custo do caminho do restaurante para cada nó de entrega
    node_costs = {}
    for node in unique_delivery_nodes:
        _, cost = a_star_path(G, start_node, node)
        node_costs[node] = cost
        
    # Ordena os nós de entrega com base no custo (tempo) do restaurante
    ordered_nodes = sorted(node_costs, key=node_costs.get)
    
    # 3. Constrói a rota completa usando A* entre os nós ordenados
    full_route = [start_node]
    total_cost = 0
    
    current_node = start_node
    for next_node in ordered_nodes:
        path, cost = a_star_path(G, current_node, next_node)
        if path:
            # Adiciona o caminho, exceto o primeiro nó (que é o current_node)
            full_route.extend(path[1:])
            total_cost += cost
            current_node = next_node
            
    # 4. Retorna ao restaurante
    path_back, cost_back = a_star_path(G, current_node, start_node)
    if path_back:
        full_route.extend(path_back[1:])
        total_cost += cost_back
        
    return full_route, total_cost, ordered_nodes

# --- 5. Visualização ---
def plot_clusters(df_deliveries, centers):
    """Gera um gráfico de dispersão dos clusters de entrega."""
    plt.figure(figsize=(10, 8))
    scatter = plt.scatter(df_deliveries['Coordenada_X'], df_deliveries['Coordenada_Y'], 
                          c=df_deliveries['Cluster'], cmap='viridis', s=50, alpha=0.7)
    plt.scatter(centers[:, 0], centers[:, 1], c='red', marker='X', s=200, label='Centroides')
    plt.title('Agrupamento de Entregas (K-Means)')
    plt.xlabel('Coordenada X')
    plt.ylabel('Coordenada Y')
    plt.legend(*scatter.legend_elements(), title="Cluster")
    plt.grid(True)
    plt.show()
